ca_tools/scripts/pos_controller_pid.py

```diff
@@ -0,0 +1,217 @@
+
+
+
+        # Computing lenght variables
+        self.error_length= self.Setpoint_length - self.Current_length
+        self.errSum_length = self.errSum_length + (self.error_length * self.timeChange)
+
+        # rospy.loginfo(Input_angle)
+        # der_error=(error - lastErr)/timeChange
+
+
+    def compute_pid_output(self):
+        # /*Compute PID Output*/
+
+        # Angle Calculations
+        if self.state=="TURN":
+            self.Output_angle = self.kp * self.error_angle + self.ki * self.errSum_angle  # + self.kd*self.der_error_angle
+            if self.Output_angle > self.Max_output_angle:
+                self.twist.angular.z = self.Max_output_angle
+            elif self.Output_angle < self.Min_output_angle:
+                self.twist.angular.z = self.Min_output_angle
+            else:
+                self.twist.angular.z = self.Output_angle
+                self.twist.linear.x=0.0
+
+        # Lenght Calculations
+        if self.state=="FORWARD":
+            self.Output_length = self.kp * self.error_length + self.ki * self.errSum_length # + self.kd*self.der_error_length
+            if self.Output_length > self.Max_output_length:
+                self.twist.linear.x = self.Max_output_length
+            elif self.Output_length < self.Min_output_length:
+                self.twist.linear.x = self.Min_output_length
+            else:
+                self.twist.linear.x = self.Output_length
+            self.twist.angular.z=0.0
+        #twist_publisher.publish(twist)
+
+
+
+    def update_time(self):
+        # lastErr=error
+        if self.now > 0:
+            self.lastTime = self.now
+
+    def set_objectives(self):
+        if self.state=="FIRST":
+            self.Input_lenght=self.Current_length
+            self.Input_angle=self.Current_angle
+            self.Setpoint_length=self.Input_lenght + 2.0
+            self.Setpoint_angle=self.Input_angle
+            self.state="FORWARD"
+            self.Set_variable=False
+            pass
+
+        if self.state=="TURN":
+            if self.Set_variable==True:
+                self.Input_lenght=self.Current_length
+                self.Input_angle=self.Current_angle
+                self.Setpoint_length=self.Input_lenght
+                self.Setpoint_angle=math.fmod(self.Input_angle + math.pi/2, math.pi*2)
+                self.Set_variable=False
+                pass
+            else:
+                pass
+
+        if self.state=="FORWARD":
+            if self.Set_variable==True:
+                self.Input_lenght=self.Current_length
+                self.Input_angle=self.Current_angle
+                self.Setpoint_length = self.Input_lenght + 2
+                self.Setpoint_angle = self.Input_angle
+                self.Set_variable=False
+            else:
+                pass
+
+    def control_objectives(self,twist_publisher):
+        #rospy.loginfo(self.error_length)
+        if self.state=="FIRST":
+            pass
+        if self.state=="FORWARD":
```

Comment:
Here you have some duplicated code too.

Reply:
It will be very hard for you to debug if you code like this. Follow the KISS principle (**keep it simple!**).